### DISCUSSION TITLE
about the loss function

### DISCUSSION THREAD

**Discussion by hunterchenghx** [ultralytics/yolov5#2569] [Q&A]
I read your loss.py carefully, and I've got questions about your build_target function.
I think you just select the target object whose bbox is related to your pre-set anchor boxes' size due to your hyp['anchor_t'], and just set these objects as your groundtruth to calculate the bbox loss, class loss and objectness loss. How about the object that is not fit with your anchors?
Also, I thought you select the nearby 3 grids for one target object, as these part of code described,
`                j, k = ((gxy % 1. < g) & (gxy > 1.)).T
                l, m = ((gxi % 1. < g) & (gxi > 1.)).T
                j = torch.stack((torch.ones_like(j), j, k, l, m))
                t = t.repeat((5, 1, 1))[j]`
but if the object is just sit on the boundary, it will duplicate the same grid (ex., it's j is True while l is also True). In that case, it will generate one more target, which will resulted in the loss. Am I correct?

**Comment by glenn-jocher** [comment]
@hunterchenghx yes that's correct!

If an object is not fit to any anchors then all of the anchor objectness truth values will be set to 0. This is why it's important to evolve good anchors, so that your Best Possible Recall (BPR) is high, and few objects are missed for training.

**Reply by hunterchenghx** [reply]
Thank you for your reply. If that, why don't you make the objects per image in training set to be at least fit to an anchor, so that all objects will be involved into the loss function. Is there any consideration? 
If only the objects that fit to the anchors are fed into the back propagation, how do you ensure after training, the network will be able to detect the objects not fit to the anchors?

**Reply by glenn-jocher** [reply]
@hunterchenghx not all output layers are suitable for detecting all objects. Layers with small receptive fields will never be able to 'see' large object and vice versa. There only objects within the anchor threshold setting will be associated with an anchor. Objects outside of this threshold will not be trained on, and will cause a reduction in Best Possible Recall. 
https://github.com/ultralytics/yolov5/blob/9f98201dd98651a768acefdd87856c86a031ff89/data/hyp.scratch.yaml#L19

**Reply by hunterchenghx** [reply]
yeah, I understand you throw away the objects outside of anchor threshold during training. For example, I tried a image which contains 6 objects. As objects are relatively small, only the 80x80 output layer's anchors are able to cover it. However, only 4 objects are selected to compute the loss. The 2 other objects are throwed away. But after training, when I use the model to do the inference, I found these 2 objects are detected correctly. I just wonder that they are not considered as the groundtruth during training, but why the trained model could detect it.

**Reply by glenn-jocher** [reply]
@hunterchenghx you're free to adapt the anchor_t hyperparameter or simply define anchors manually in your model.yaml file to suit your customization requirements.
